pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -0,0 +1,219 @@
+	return ttu
+}
+
+// CheckTimeout is called with a token when it is seen by an authenticator
+// it touches only the tokenChannel so it is safe to call from other threads
+func (a *TokenTimeoutUpdater) CheckTimeout(name string, token *oauth.OAuthAccessToken) error {
+	var timeout time.Time
+	timenow := time.Now()
+	if token.TimeoutsIn > 0 {
+		timeout = token.CreationTimestamp.Time.Add(time.Duration(token.TimeoutsIn) * time.Second)
+		if timeout.Before(timenow) {
+			return ErrTimedout
+		}
+	}
+
+	// After a positive timeout check we need to update the timeout and
+	// schedule an update so that we can either set or update the Timeout
+	a.tokenChannel <- tokenData{name, token.ClientName, timenow, timeout}
+	return nil
+}
+
+// updates known clients timeouts, removing clients if they go away
+// and calculate the flush period based on actual timeouts, with a
+// 10 minutes default
+func (a *TokenTimeoutUpdater) updateClientTimeouts() time.Duration {
+	// Default to 10 minutes
+	var flush int32 = 600
+
+	// Update known ones only
+	if len(a.clientTimeouts) > 0 {
+		for name := range a.clientTimeouts {
+			c, err := a.oauthClient.Get(name, metav1.GetOptions{})
+			if err != nil {
+				if kerrors.IsNotFound(err) {
+					delete(a.clientTimeouts, name)
+				}
+			}
+			if c.AccessTokenTimeoutSeconds == nil {
+				delete(a.clientTimeouts, name)
+			} else {
+				a.clientTimeouts[name] = *c.AccessTokenTimeoutSeconds
+				// Use half period if it is smaller
+				half := int32(math.Ceil(float64(a.clientTimeouts[name]) / 2))
+				if flush > half {
+					flush = half
+				}
+			}
+		}
+	}
+	return time.Duration(flush) * time.Second
+}
+
+func (a *TokenTimeoutUpdater) clientTimeout(name string) int32 {
+	t, ok := a.clientTimeouts[name]
+	if !ok {
+		c, err := a.oauthClient.Get(name, metav1.GetOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				glog.Info("Token timeout set but OAuthClient %s not found!", name)
+			} else {
+				glog.Error("Failed to fetch OAuthClient for timeout value")
+			}
+			return 0
+		}
+		if c.AccessTokenTimeoutSeconds == nil {
+			glog.Info("Token timeout set but OAuthClient %s has no timeout!", name)
+			return 0
+		}
+		t = *c.AccessTokenTimeoutSeconds
+		a.clientTimeouts[name] = t
+	}
+	return t
+}
+
+func (a *TokenTimeoutUpdater) insert(td tokenData) {
+	a.data[td.name] = td
+	a.tree.ReplaceOrInsert(&tokenDataRef{td.name, td.timeout})
+}
+
+func (a *TokenTimeoutUpdater) remove(td tokenData) {
+	a.tree.Delete(&tokenDataRef{td.name, td.timeout})
+	delete(a.data, td.name)
+}
+
+func (a *TokenTimeoutUpdater) flush(flushHorizon time.Time) {
+	flushedTokens := 0
+	totalTokens := len(a.data)
+
+	glog.Infof("Flushing tokens timing out before %v", flushHorizon)
+
+	for {
+		item := a.tree.Min()
+		if item == nil {
+			// out of items
+			break
+		}
+		td := a.data[item.(*tokenDataRef).name]
+		if td.timeout.Before(flushHorizon) {
+			delta := a.clientTimeout(td.client)
+			// ignore tokens for which we can't find a delta
```

Comment:
Doesn't this cause tokens that should never expire to expire now?

Reply:
A token that should never expire simply should never have a TimeoutsIn value, so they are never inserted in the system.
If they are and the timeout vanishes from the oauth clients, I figured we may want to timeout the current token, so that the user will get a new one that has no TimeoutsIn value.
We can change this to erase the TimeoutsIn value from the token, but then we have to be absolutely sure the OauthClient was changed (not that the call to figure out its value has changed). I may have that confidence with the Informer for OAuth Clients, in that case I might change this code to modify the client to remove the timeout.
It's unlikely this is going to be an interesting case and it will be super rare anyway.

Reply:
I changed this to set the timeoutsin value to 0